### PR TITLE
fix(wasm): validate filters in all persistence modes

### PR DESCRIPTION
**Pull request by flrgh**
### Summary

The primary purpose of this change is to ensure that each `.filters[].name` is validated against available/installed filters in dbless mode (this was only working in db mode prior).

I also added an additional entity check which will return a more descriptive error when wasm is disabled or no filters are present (this is more useful for dbless mode than anything).

### Checklist

- [x] The Pull Request has tests
- ~[ ] There's an entry in the CHANGELOG~
- ~[ ] There is a user-facing docs PR~

### Issue reference

KAG-2059

### CODE REVIEW

**Review by hishamhm — COMMENTED** [review]
Looks like a more consistent approach overall!

File: kong/db/schema/entities/filter_chains.lua [hunk 1]
```diff
@@ -65,5 +68,41 @@ return {
...
+    -- ```
+    -- name = "expected one of: "
+    -- ```
+    --
+    -- This check simply adds an entity-level error message in those cases to
+    -- clarify.
+    { custom_entity_check = {
+        field_sources = { "filters" },
+        run_with_invalid_fields = true,
+        fn = function(_)
```
[the remark this answers is not in the document]
Comment (flrgh) [reply kong/db/schema/entities/filter_chains.lua:88]: I like this, good idea.

File: spec/02-integration/20-wasm/02-db_spec.lua [hunk 2]
```diff
@@ -314,8 +317,8 @@ describe("wasm DB entities [#" .. strategy .. "]", function()
           assert.is_table(err_t.fields)
           assert.same({
             filters = {
-              [2] = { name = "no such filter: missing" },
-              [4] = { name = "no such filter: also-missing" },
+              [2] = { name = "expected one of: test, other" },
+              [4] = { name = "expected one of: test, other" },
```
[the remark this answers is not in the document]
Comment (flrgh) [reply spec/02-integration/20-wasm/02-db_spec.lua:321]: Honestly I think it's unwieldy for _any_ `one_of` that isn't a static list, but I didn't know it was possible to override the error string. I'll give this a try!

File: kong/db/schema/entities/filter_chains.lua (hunk 1 above)
[the remark this answers is not in the document]
Comment (flrgh) [reply kong/db/schema/entities/filter_chains.lua]: I've implemented your suggestions around this and the other `kong.runloop.wasm()` Lua API bits. This lead to a bit more refactoring than I planed on, so I'm not 100% sold on the result, but I think it's plenty good for the time being.

Please give the newest commit a look and let me know what ya think!

[the remark this answers is not in the document]
Comment (hishamhm) [reply kong/db/schema/entities/filter_chains.lua]: > This lead to a bit more refactoring than I planed on, so I'm not 100% sold on the result, but I think it's plenty good for the time being.

It is! The `enable` vs `_M.enable` bit had me do a double-take (unfortunately no other naming suggestion comes to mind), but overall it looks pretty nice!

File: kong/runloop/wasm.lua [hunk 3]
```diff
@@ -48,6 +72,8 @@ local TYPE_COMBINED = 2
 
 
 local ENABLED = false
+local STATUS = "wasm support is not enabled"
```
Comment (zhongweiy) [inline kong/runloop/wasm.lua:75]: nit: "wasm support is not enabled" seems has been repeated several times in this file. I'm thinking whether we will have const variable to store this string. But it seems `const` is only available in Lua 5.4.

File: spec/02-integration/20-wasm/08-declarative_spec.lua [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+        break
+      end
+    end
+  end
+
+  assert.is_true(found, "expected '" .. err .. "' message in response")
+end
+
+
+describe("#wasm declarative config", function()
```
Comment (zhongweiy) [inline spec/02-integration/20-wasm/08-declarative_spec.lua:50]: This `08-declarative_spec.lua` only contains negative test cases. Shall we add a positive test case like the filter is successfully added for db-less? Or it already has been included in other test file?

File: kong/runloop/wasm.lua (hunk 3 above)
Comment (hishamhm) [reply kong/runloop/wasm.lua:75]: while we're at it, I wasn't going to comment on style preferences, but... (—the best way to start a review comment, right? :laughing: ) at first glance the all-uppercase variables do look like they're intended as constants. If they're meant as internal status-keeping variables, perhaps something like `local wasm_enabled` and `local wasm_status` might be easier to understand at a glance (and then you could do the likes of `local wasm_status = WASM_STATUS_NOT_ENABLED` if you wanted to centralize the messages at the top). Just a thought!

File: spec/02-integration/20-wasm/08-declarative_spec.lua (hunk 4 above)
Comment (flrgh) [reply spec/02-integration/20-wasm/08-declarative_spec.lua:50]: This is a good point. `05-cache-invalidation_spec.lua` _does_ test the same dbless API endpoint (`/config`), but it doesn't hurt to have at least a simple happy-path example in this file.

Comment (zhongweiy) [reply spec/02-integration/20-wasm/08-declarative_spec.lua:50]: Thanks for pointing out. Look good to me then.

File: kong/runloop/wasm.lua (hunk 3 above)
Comment (flrgh) [reply kong/runloop/wasm.lua]: That's fair. I've moved the status strings into local, "constant"-y `STATUS_*` vars at the module level.

**Review by zhongweiy — APPROVED** [review]

File: kong/runloop/wasm.lua (hunk 3 above)
Comment (flrgh) [reply kong/runloop/wasm.lua]: Ugh, yeah I wish Lua were better about this. I like my `SCREAMING_SNAKE_CASE` for constants, but I _also_ like it for "this variable is from the outermost scope in the file"

**Pull request merged**